Solver Testing and Addressed Bugs Along the Way

vincent-maillou opened the pull request:
Added testing suite for sequential and distributed solvers. Tests are separated in two parts: `unit/` and `component_integration/`.
- Fixed the factorization in scipy solver (faked Cholesky -> LU)
- Fixed problems in distributed structured solver RHS (solve / mapping / gather)
- Added "automated" runner script. To be seen with GitHub action if we keep this, if it's needed or not.

Regarding #129 , the tests are passing for CPU/GPU arrays using MPI (not tested with `mpi` `cuda` aware and NCCL, need environment development from branch `installation_environements`), it might still be that the exact case that fail is not tested for and / or I have overlooked something.

Closes #72 , as now the testing suite structure will be defined and about 300–500 tests are available. A lot of work remaining to test the entire framework → will be done incrementally

The PR:  Mpi issues #124 has been merged in subsequent "AR1/gamma_prior" branches, so this is gonna be dealt with in rebasing to these branches. For this reason, I would suggest closing `Mpi issues #124`

vincent-maillou committed:
BUG,API: The sparse solver wasn't correct. Faking Cholesky using
a non-permuted LU factorization isn't actually numerically correct.
Modified the implementation (and interfaces of the solvers) to use
LU and the appropriate triangular solve calls. Other solvers still relies
on Cholesky when available.

vincent-maillou committed:
TST: First coherent coponent_integration tetsing structure and firsts tests for Serinv.
- Figured out fixture structure
- Added fixtures factory for testing utils functions
- Added structured solver testing for solver.factorize workflow

vincent-maillou committed:
TST, BUG: fix factorize tests (that werent' asserting correctness)
- This raised an unexpected behavior in the `_strutured_to_spmatrix` mapping.

vincent-maillou committed:
TST: All tests for sequential structured_solver

vincent-maillou committed:
TST: Added sequential sparse solver tests (when applicable)

vincent-maillou committed:
TST, STY: Added tests for dense solvers and formated test suite

vincent-maillou committed:
TST: Added support for GPU testing + runner script with automatic backend detection

vincent-maillou committed:
TST: Unified testing configs across testing suite

vincent-maillou committed:
TST: Used `tests` package level tolerances

vincent-maillou committed:
MAINT: Renamed tests folders

vincent-maillou committed:
MAINT: Refactored out the structured solvers testing utils

vincent-maillou committed:
TST: Added unit tests for structured solvers mapping kernels

vincent-maillou committed:
TST, MAINT: Re-ordered testing config/fixtures
- Separated sequential / distributed tests related fixtures

vincent-maillou committed:
TST, BUG, MAINT: Added tests dist serinv mapping
- Solved several issues in the implementation of the distributed solver
- Fixed the structured_to_sparse mapping function that simply wasn't working

vincent-maillou committed:
STY: Formatting

vincent-maillou committed:
TST: Added tests for distributed structured solver
- Adapted `runner.sh`
- Updated `README.md` with new test descriptions

vincent-maillou committed:
TST, BUG: Fixed runner script error with mpi detection

vincent-maillou committed:
MAINT: Fixed typos

vincent-maillou committed:
MAINT: Copilot correcting my typos...

lisa-gm commented on src/dalia/solvers/distributed_structured_solver.py:
```diff
@@ -763,16 +793,17 @@ def _structured_to_spmatrix(
 
         # TODO: Need to communicate to agregates/Map the local B matrix to all ranks
```
someone once told me i shouldnt have any TODOs left in my PRs.

lisa-gm commented on src/dalia/solvers/structured_solver.py:
what are your thoughts on this "symmetrize" option? do you want it to be enabled by default? when is it relevant / not relevant?

vincent-maillou replied:
For now the default is to symmetrize. Since this was matching the expected behavior from DALIA is felt more logic to do it this way to the option doesn't have to appear everywhere in the code.
In the tests this is made explicit.
I don't have further opinions for now. I would leave it as is it now.

lisa-gm commented on tests/component_integration/solvers/dense_solvers/utils.py:
after the discussions of the last week, i wonder whether we should have a non-diagonally dominant test case?

lisa-gm commented on tests/component_integration/solvers/sparse_solvers/sequential/test_factorize.py:
can't you test the solver by checking if L is lower triangular and L*L^T = A?

lisa-gm commented on tests/component_integration/solvers/utils.py:
```diff
@@ -0,0 +1,182 @@
+
+
+def _reference_cholesky(A):
+    """Compute reference Cholesky decomposition using NumPy.
```
technically numpy / cupy? holds for below ones as well ...

vincent-maillou replied:
Thanks! I adjusted the doc here and in other places of the test-suite to accommodate that.

vincent-maillou committed:
TST, MAINT: Addressed comments in PR
- Removed DiagDom caracteristic in dense solver testing
- Minor modifications in the sparse solver implementation
- Improved test for the scipy sparse solver

vincent-maillou committed:
MAINT: Ran formatting tools
